### PR TITLE
Add server implementation

### PR DESCRIPTION
**Pull request by foxcpp**
Note: I renamed field name and adjusted test cases 
so I could use them to verify formatting too.

### CODE REVIEW

**Comment by foxcpp** [comment]
In regards of test changes. TestThreadParsing passes test.thread to parseThreadResp. **But**, it is not a single thread, it is a complete response. It is a subtle difference so I renamed the field along the way to better match what is going on.

I also wrapped strings in imap.RawString as this is what formatting code generates. This in turn required a small change in parsing code to make it accept RawString.

**Comment by foxcpp** [comment]
Additionally, I believe there is a problem in the original first test case.
```
[]interface{}{"1", "2", "3", "4"}
```
corresponds to
```
* THREAD 1 2 3 4
```
which is not a well-formed response per ABNF:
```
thread-data     = "THREAD" [SP 1*thread-list]
thread-list     = "(" (thread-members / thread-nested) ")"
```
@keur, is it intended that parser is forgiving and interprets this as a single thread? Note that it is also meaningful to interpret such output as if it was `* THREAD (1) (2) (3) (4)`. By being forgiving here one risks poisoning ecosystem and breaking interoperability.

Footnote: Correspondence of []interface{} to IMAP wire format can be easily observed by inserting following into test case:
```go
b := strings.Builder{}
imap.NewUntaggedResp(test.response).WriteTo(imap.NewWriter(&b))
t.Log(b.String())
t.Log(test.str)
```

I took the liberty to update test case so my own test based on same cases would pass.

**Comment by foxcpp** [comment]
~~There is bug in command parsing that makes search criteria always nil.~~

**Comment by kkuehlz** [comment]
On Mon, 20 Jul 2020 07:20:18 -0700
Max Mazurov <notifications@github.com> wrote:

> That is, I believe there is a problem in the original first test case.
> ```
> []interface{}{"1", "2", "3", "4"}
> ```
> corresponds to
> ```
> * THREAD 1 2 3 4
> ```
> which is not a well-formed response per ABNF:
> ```
> thread-data     = "THREAD" [SP 1*thread-list]
> thread-list     = "(" (thread-members / thread-nested) ")"
> ```
> @keur, is it intended that parser is forgiving and interprets this as
> a single thread?

No, I wasn't trying to deliberately trying to break ABNF. It's been a
while since I implemented this, but I probably intended to recreate
THREAD (1) (2) (3) (4) and got messed up by all the []interface{}.

**Comment by foxcpp** [comment]
It annoys me that go-imap-sortthread is 1.0 already and I would really like to fold ESORT support into it. So, @emersion, I would prefer you to not do a release after merging this PR and give me time to work on ESORT.

**Comment by foxcpp** [comment]
Or I might try to add ESORT for server as part of this PR.

**Comment by emersion** [comment]
>It annoys me that go-imap-sortthread is 1.0 already and I would really like to fold ESORT support into it. So, @emersion, I would prefer you to not do a release after merging this PR and give me time to work on ESORT.

Why does this matter? Can't we just e.g. tag 1.1 with this PR, and 1.2 with ESORT?

File: thread_test.go [hunk 1]
```diff
@@ -141,7 +149,12 @@ var threadTests = []struct {
 func TestThreadParsing(t *testing.T) {
 	for _, test := range threadTests {
 		t.Run(test.name, func(t *testing.T) {
-			threads, err := parseThreadResp(test.thread)
+			b := strings.Builder{}
```
Comment (emersion) [inline thread_test.go]: Remnant from debugging session?

File: commands.go [hunk 2]
```diff
@@ -26,8 +28,72 @@ func (cmd *SortCommand) Command() *imap.Command {
 	}
 }
 
+func parseSortCrit(fields interface{}) ([]SortCriterion, error) {
```
Comment (emersion) [inline commands.go]: Let's not use an abbreviation here: `parseSortCriteria` isn't too long.

**Comment by foxcpp** [comment]
> Why does this matter? Can't we just e.g. tag 1.1 with this PR, and 1.2 with ESORT?

ESORT extends Sort command and in my opinion the most clean backend interface change is to modify Sort method added by this PR (since it pretty much just adds an argument and changes return value slightly). 

Probably we can go with Sort and ExtendedSort methods, if you want.

File: commands.go [hunk 3]
```diff
@@ -26,8 +28,72 @@ func (cmd *SortCommand) Command() *imap.Command {
...
+	if reverse {
+		return nil, errors.New("Missing sort key after REVERSE")
+	}
+
+	return result, nil
+}
+
 func (cmd *SortCommand) Parse(fields []interface{}) error {
-	return errors.New("sortthread: not yet implemented")
+	if len(fields) < 3 {
```
Comment (emersion) [inline commands.go:70]: Is a command with 2 arguments valid?

File: server.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import (
+	"errors"
+
+	"github.com/emersion/go-imap"
+	"github.com/emersion/go-imap/server"
+)
+
+var ErrUnsupportedBackend = errors.New("sortthread: backend not supported")
+
+type SortMailbox interface {
```
Comment (emersion) [inline server.go:13]: Should we embed `imap.Mailbox` here?

File: commands.go (hunk 3 above)
Comment (foxcpp) [reply commands.go:70]:
```
BASE.6.4.SORT. SORT Command

   Arguments:  sort program
               charset specification
               searching criteria (one or more)
```

No, unlike SEARCH, charset argument is mandatory.

File: thread_test.go (hunk 1 above)
Comment (foxcpp) [reply thread_test.go]: Oops, yeah.

File: commands.go (hunk 2 above)
Comment (foxcpp) [reply commands.go]: Still curing my desire to abbreviate names :)

File: server.go (hunk 4 above)
Comment (foxcpp) [reply server.go:13]: backend.Mailbox, yes, we should as I already answered in "Merge small extensions" PR.

File: commands.go (hunk 3 above)
Comment (emersion) [reply commands.go:70]: OK

File: commands.go [hunk 5]
```diff
@@ -47,3 +113,31 @@ func (cmd *ThreadCommand) Command() *imap.Command {
 		},
 	}
 }
+
+func (cmd *ThreadCommand) Parse(fields []interface{}) error {
+	if len(fields) < 2 {
```
Comment (emersion) [inline commands.go]:
```
BASE.6.4.THREAD. THREAD Command

Arguments:  threading algorithm
            charset specification
            searching criteria (one or more)
```

So this should be `< 3` maybe?

Comment (foxcpp) [reply commands.go]: Yes, you are right.

File: server.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	backend.Mailbox
+	Sort(uid bool, sortCrit []SortCriterion, searchCrit *imap.SearchCriteria) ([]uint32, error)
+}
+
+type ThreadBackend interface {
+	SupportedThreadAlgorithms() []ThreadAlgorithm
```
Comment (emersion) [inline server.go:20]: We should embed `backend.Backend` here.

Comment (foxcpp) [reply server.go:20]: Done.

**Comment by foxcpp** [comment]
> > Why does this matter? Can't we just e.g. tag 1.1 with this PR, and 1.2 with ESORT?
> 
> ESORT extends Sort command and in my opinion the most clean backend interface change is to modify Sort method added by this PR (since it pretty much just adds an argument and changes return value slightly).

... this will also encourage backend developers to implement both.

**Comment by emersion** [comment]
>ESORT extends Sort command and in my opinion the most clean backend interface change is to modify Sort method added by this PR (since it pretty much just adds an argument and changes return value slightly).

OK, works for me

**Review by emersion — APPROVED** [review]
LGTM, thanks!

**Pull request merged**